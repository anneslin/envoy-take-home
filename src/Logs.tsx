import React, { useEffect, useState } from 'react';
import EntryType from './entry';
import { KEY, SIGN_OUT_URL } from './constants';
import moment from 'moment';

enum Filters {
  none = 0,
  signedOut = 1,
  inOffice = 2
};

interface LogsProps {
  searchValue: string;
  entries: ReadonlyArray<EntryType>;
  fetchEntries: () => void;
};

const Logs = ({ searchValue, entries, fetchEntries }: LogsProps) => {
  const [signingOutIds, setSigningOutIds] = useState([] as ReadonlyArray<string>);
  const [displayEntries, setDisplayEntries] = useState(entries);
  const [filter, setFilter] = useState(Filters.none);

  useEffect(() => setDisplayEntries(entries), [entries]);
  // very basic search + filter filtering
  useEffect(() => {
    if (searchValue !== "") {
      const s = searchValue.toLowerCase()
      const searchResults = entries.filter(x =>
        (x.name.toLowerCase().includes(s)
        || x.notes.toLowerCase().includes(s)
        || (x.signOut && x.signOut.toLowerCase().includes(s)))
        && (filter === Filters.none || (filter === Filters.signedOut ? x.signOut : !x.signOut)));
      setDisplayEntries(searchResults);
    } else {
      setDisplayEntries(entries.filter(x => (filter === Filters.none || (filter === Filters.signedOut ? x.signOut : !x.signOut))));
    }
  }, [searchValue, entries, filter]);

  const toggleFilter = (f: Filters) => f === filter ? setFilter(Filters.none) : setFilter(f);

  const signOut = (id: string, idx: number) => {
    setSigningOutIds(signingOutIds.concat(id));
    const data = { data: { type: "entries", id: id } }
    fetch(SIGN_OUT_URL, {
      method: "POST",
      body: JSON.stringify(data),
      headers: {
        'X-Api-Key': KEY,
        'Content-Type': 'application/json'
      }
    })
      .then((response: Response) => {
        if (!response.ok) {
          setSigningOutIds(signingOutIds.filter(x => x !== id));
          throw Error(response.statusText);
        }
        fetchEntries();

      }).catch((error) => {
        console.log(error);
    });
  };

  return (
    <div className="flex-grow h-screen overflow-y-scroll">
      <div className="mx-auto">
        <div className="mt-8">
          <div className="flex items-center">
            <div className="filterText">Filter By:</div>
            <div className={filter === Filters.signedOut ? 'filter selected' : 'filter'} onClick={() => toggleFilter(Filters.signedOut)}>Signed Out</div>
            <div className="filterText">|</div>
            <div className={filter === Filters.inOffice ? 'filter selected' : 'filter'} onClick={() => toggleFilter(Filters.inOffice)}>In Office</div>
          </div>
          <table className="w-full">
            <thead>
              <tr>
                <th className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest third">Name</th>
                <th className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest third">Notes</th>
                <th className="text-sm font-semibold text-grey-darker p-1 bg-grey-lightest third">Signed out</th>
              </tr>
            </thead>
            <tbody className="align-baseline">
              {displayEntries.map((x, i) =>
                <tr key={x.id}>
                  <td className="p-2 border-t border-grey-light font-mono text-xs">{x.name}</td>
                  <td className="p-2 border-t border-grey-light font-mono text-xs">{x.notes}</td>
                  <td className="p-1 border-t border-grey-light font-mono text-xs">
                    {x.signOut && moment(x.signOut).format('lll')}
                    {x.signOut == null && (signingOutIds.includes(x.id)
                      ? <button className="btn disabled btn--smaller btn--outline">
                        Signing out <i className="fas fa-spinner"></i>
                      </button>
                      : <button className="btn btn--smaller btn--outline" onClick={() => signOut(x.id, i)}>
                        Sign out
                        </button>
                    )}

                  </td>
                </tr>
              )}
            </tbody>
          </table>
          {displayEntries.length === 0 &&
            <div className="p-2 border-t border-grey-light font-mono text-xs text-center">No Entries{searchValue !== "" && ` for "${searchValue}"`}</div>
          }

          {/* <div className="mt-8 mb-8">
            <p className="mt-2">
              <div className="flex">
                <a href="#" className="mr-4 px-1">
                  <i className="fas fa-angle-left" />
                </a>
                <div className="flex text-center mr-4">
                  <span className="w-20">
                    1
              </span>
                  <span className="w-20">&nbsp;/&nbsp;</span>
                  <span className="w-20">
                    2
              </span>
                </div>
                <a href="#" className="mr-4 px-1">
                  <i className="fas fa-angle-right" />
                </a>
              </div>
            </p>
          </div> */}

        </div>
      </div>
    </div>
  );
}

export default Logs; 