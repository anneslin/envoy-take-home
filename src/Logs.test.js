import React from 'react';
import ReactDOM from 'react-dom';
import Logs from './Logs';
import TestRenderer from 'react-test-renderer';

it('renders Logs without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<Logs searchValue={""} entries={[]} fetchEntries={() => {}}/>, div);
  ReactDOM.unmountComponentAtNode(div);
});

const entries = [
  {id: "1231", name: "Johnny Duh", notes:"Hello you!", signOut: null},
  {id: "3333", name: "Helena Montana", notes:"Who?", signOut: null},
  {id: "3213", name: "Mable Able", notes:"YEs I CaN!", signOut: "2020-03-09T06:25:58.313Z"}
];

test('renders all entries', () => {
  const tree = TestRenderer
    .create(<Logs searchValue={""} entries={entries} fetchEntries={() => {}}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test('renders only Helena Montana with searchValue', () => {
  const tree = TestRenderer
    .create(<Logs searchValue={"helena"} entries={entries} fetchEntries={() => {}}/>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
