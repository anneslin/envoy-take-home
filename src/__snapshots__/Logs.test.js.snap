// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders all entries 1`] = `
<div
  className="flex-grow h-screen overflow-y-scroll"
>
  <div
    className="mx-auto"
  >
    <div
      className="mt-8"
    >
      <div
        className="flex items-center"
      >
        <div
          className="filterText"
        >
          Filter By:
        </div>
        <div
          className="filter"
          onClick={[Function]}
        >
          Signed Out
        </div>
        <div
          className="filterText"
        >
          |
        </div>
        <div
          className="filter"
          onClick={[Function]}
        >
          In Office
        </div>
      </div>
      <table
        className="w-full"
      >
        <thead>
          <tr>
            <th
              className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest third"
            >
              Name
            </th>
            <th
              className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest third"
            >
              Notes
            </th>
            <th
              className="text-sm font-semibold text-grey-darker p-1 bg-grey-lightest third"
            >
              Signed out
            </th>
          </tr>
        </thead>
        <tbody
          className="align-baseline"
        >
          <tr>
            <td
              className="p-2 border-t border-grey-light font-mono text-xs"
            >
              Johnny Duh
            </td>
            <td
              className="p-2 border-t border-grey-light font-mono text-xs"
            >
              Hello you!
            </td>
            <td
              className="p-1 border-t border-grey-light font-mono text-xs"
            >
              <button
                className="btn btn--smaller btn--outline"
                onClick={[Function]}
              >
                Sign out
              </button>
            </td>
          </tr>
          <tr>
            <td
              className="p-2 border-t border-grey-light font-mono text-xs"
            >
              Helena Montana
            </td>
            <td
              className="p-2 border-t border-grey-light font-mono text-xs"
            >
              Who?
            </td>
            <td
              className="p-1 border-t border-grey-light font-mono text-xs"
            >
              <button
                className="btn btn--smaller btn--outline"
                onClick={[Function]}
              >
                Sign out
              </button>
            </td>
          </tr>
          <tr>
            <td
              className="p-2 border-t border-grey-light font-mono text-xs"
            >
              Mable Able
            </td>
            <td
              className="p-2 border-t border-grey-light font-mono text-xs"
            >
              YEs I CaN!
            </td>
            <td
              className="p-1 border-t border-grey-light font-mono text-xs"
            >
              Mar 9, 2020 2:25 AM
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;

exports[`renders only Helena Montana with searchValue 1`] = `
<div
  className="flex-grow h-screen overflow-y-scroll"
>
  <div
    className="mx-auto"
  >
    <div
      className="mt-8"
    >
      <div
        className="flex items-center"
      >
        <div
          className="filterText"
        >
          Filter By:
        </div>
        <div
          className="filter"
          onClick={[Function]}
        >
          Signed Out
        </div>
        <div
          className="filterText"
        >
          |
        </div>
        <div
          className="filter"
          onClick={[Function]}
        >
          In Office
        </div>
      </div>
      <table
        className="w-full"
      >
        <thead>
          <tr>
            <th
              className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest third"
            >
              Name
            </th>
            <th
              className="text-sm font-semibold text-grey-darker p-2 bg-grey-lightest third"
            >
              Notes
            </th>
            <th
              className="text-sm font-semibold text-grey-darker p-1 bg-grey-lightest third"
            >
              Signed out
            </th>
          </tr>
        </thead>
        <tbody
          className="align-baseline"
        >
          <tr>
            <td
              className="p-2 border-t border-grey-light font-mono text-xs"
            >
              Helena Montana
            </td>
            <td
              className="p-2 border-t border-grey-light font-mono text-xs"
            >
              Who?
            </td>
            <td
              className="p-1 border-t border-grey-light font-mono text-xs"
            >
              <button
                className="btn btn--smaller btn--outline"
                onClick={[Function]}
              >
                Sign out
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
